spec/models/relation_spec.rb

```diff
@@ -38,17 +38,24 @@
   let(:relation) { build(:relation, from:, to:, relation_type: type) }
 
   it "validates lag numericality" do
+    relation.lag = nil
+    expect(relation).to be_valid
     relation.lag = -1
+    expect(relation).to be_valid
+    relation.lag = 2_000
+    expect(relation).to be_valid
+    relation.lag = -2_000
+    expect(relation).to be_valid
+
+    relation.lag = 2_001
     expect(relation).not_to be_valid
     expect(relation.errors[:lag])
-      .to include(I18n.t(:"activerecord.errors.models.relation.attributes.lag.greater_than_or_equal_to.zero"))
+      .to eq(["must be less than or equal to #{Relation::MAX_LAG}."])
 
-    relation.lag = 2_147_483_648
+    relation.lag = -2_001
     expect(relation).not_to be_valid
-    expect(relation.errors[:lag]).to include(I18n.t(:"activerecord.errors.models.relation.attributes.lag.less_than_or_equal_to"))
-
-    relation.lag = 1_000
-    expect(relation).to be_valid
+    expect(relation.errors[:lag])
+      .to eq(["must be greater than or equal to #{Relation::MIN_LAG}."])
```

Comment:
I personally find [Shoulda Matchers](https://matchers.shoulda.io/) quite helpful when testing AR validations. You could use the [`validate_numericality_of`](https://matchers.shoulda.io/docs/v6.5.0/Shoulda/Matchers/ActiveModel.html#validate_numericality_of-instance_method) or [`allow_value`](https://matchers.shoulda.io/docs/v6.5.0/Shoulda/Matchers/ActiveModel.html#allow_value-instance_method) matchers.

```suggestion
   expect(relation).to validate_numericality_of(:lag)
     .is_greater_than_or_equal_to(-2_000)
     .is_less_than_or_equal_to(2_000)
     .allow_nil
```

Reply:
True, but I always felt a bit hesitant with these shoulda matchers, because they really just seem to duplicate what's in the model, just written another way:

```ruby
   expect(relation).to validate_numericality_of(:lag)
     .is_greater_than_or_equal_to(-2_000)
     .is_less_than_or_equal_to(2_000)
     .allow_nil
```
vs
```ruby
  validates :lag, numericality: {
    allow_nil: true,
    less_than_or_equal_to: 2_000,
    greater_than_or_equal_to: -2_000
  }
```

That's the same code.
The test only ensures that this validation is never deleted, and creates a maintenance burden whenever the values in the validator are changed.

To be honest I was hesitant to just delete that test. It was doing more before because the i18n was different, so it kinda made sense, but now I don't feel it's that useful.

Nevertheless, if we are to keep that test, then using should matchers is more expressive. I'll use your proposition.

Reply:
The failure message with shoulda_matchers is particularly nice:

```
  1) Relation validates lag numericality
     Failure/Error:
       expect(relation).to validate_numericality_of(:lag)
         .is_greater_than(Relation::MIN_LAG)
         .is_less_than_or_equal_to(Relation::MAX_LAG)
         .allow_nil

       Expected Relation to validate that :lag looks like a number greater than
       -2000 and less than or equal to 2000 as long as it is not nil, but this
       could not be proved.
         In checking that Relation disallows :lag from being a number that is
         not greater than -2000, after setting :lag to ‹-2000›, the matcher
         expected the Relation to be invalid, but it was valid instead.
```

Reply:
And actually this test will be useful when we'll restrict the lag property to relations of type follows/precedes. It makes no sense to allow setting lag for a relation of type relates_to (for instance). Jens created a bug ticket a while ago about that.